adding roadmap and changelog docs

tsurdilo opened the pull request:
This adds roadmap and changelog docs. 

Its a draft so need everyones inputs please before adding it. 
The entries here are just what I could come up with so nothing is final or determined. Please
add your inputs so we can finalize. 

Once merged I would like to ask if we can take turns updating this documents. Maybe once per month would be enough. 
This will also drive better communication (discussion please over slack channel) and show community our progress.

salaboy commented on workflow/spec/roadmap.md:
```diff
@@ -0,0 +1,40 @@
+* Update sections to make their contents clearer to understand
+* Update images
+
+### Finalize serverless workflow Notation
```
@tsurdilo I would like to see here more iterations on the notation side (diagram), can we add an item for comparing with other existing frameworks/notations? So we can align with other notations that are well-known already?

salaboy replied:
I'm interested in comparing the proposed notation to existing workflow notations such as AWS Step Functions, BPMN and BPEL as well.. I can work on creating a table comparing these graphical notations so we can compare and make sure that we are clear on the direction

tsurdilo replied:
updated for notation

salaboy commented on workflow/spec/roadmap.md:
```diff
@@ -0,0 +1,40 @@
+* Add a command-line linter for serverless workflow JSON.
+
+### Provide serverless workflow Fluent Java API
+* Add a fluent Java API/SPI. More languages would be great to have as contributions as well.
```
@tsurdilo can we remove "Java" from here? maybe what we can do is provide an spec for how the API should look like so we can provide different implementations.

tsurdilo replied:
updated

salaboy commented on workflow/spec/roadmap.md:
```diff
@@ -0,0 +1,40 @@
+Serverless Workflow specification is a community effort so any feedback is more than welcome!
+The best ways to provide feedback is contact us (see the "Communication" section in the [readme doc](readme.md))
+or open an Issue in this GitHub project. You can also open a Pull Request to directly 
+add your feedback to this document. 
```
@tsurdilo and all.. I think that we need a place to describe what is the "value added" and the "scope" for this specification. By not having that stated clearly it is very difficult to justify the adoption of a new spec over other well know specifications and proprietary notations that are already implemented by different providers.

tsurdilo replied:
would it help if entries in roadmap were in a table format? what would you like to see as column names?

duglin replied:
One thing we've done for CloudEvents is create a primer: https://github.com/cloudevents/spec/blob/master/primer.md   that provides insight and in-depth explanations behind the reasons for the spec and some of the design decisions we made. Stuff that isn't really appropriate for a formal specification but useful for people who like to understand the rationale behind the decisions made. Asking people to read github issues/PR to get the background isn't as easy to do.  Then in a doc like that you can position this spec against others to explain why this is different and needed.

tsurdilo replied:
@duglin good idea. will work on that

salaboy replied:
@tsurdilo is there any work on that (primer).. I think that it is a fundamental piece of information

duglin commented on workflow/spec/changelog.md:
```diff
@@ -0,0 +1,22 @@
+# Serverless Workflow - Changelog
```
You might find it easier to just run a git cmd to auto-generate this list on a periodic basis (and remove trivial ones). Just because I know how I tend to work... I always forget to update things like this :-)  
But aside from that it seems fine.

duglin commented on workflow/spec/roadmap.md:
```diff
@@ -0,0 +1,42 @@
+* Add comparisons and align with other existing frameworks/notations
+
+### Add JSON examples
+* Currently JSON Schema definition is provided for each of the workflow elements. Add JSON example for each aswell.
```
s/aswell/as well/   Unless you meant "a swell" ? :-)

tsurdilo replied:
updated

duglin commented on workflow/spec/roadmap.md:
```diff
@@ -0,0 +1,42 @@
+### Create content for community
+* Create blogs/videos showcasing serverless workflow features
+
+## Longer Term
```
I wonder if the word "Options" should be added here.  I say this because of the last one... the "workflow API". Does everyone agree with that as something that will be done?  If not it might be good to make it clear that this list is just ideas people have and not necessarily a statement of commitment.  Unless you've agreed.... then never mind.

tsurdilo replied:
@duglin we have a java based api/spi already which would like to contribute to cncf (https://github.com/tsurdilo/workflow-api), so this already exists just probably was waiting until we get a new repo where it probably makes more sense to live. wdyt.

duglin replied:
sounds great to me - especially if we could get other languages supported too... like golang

duglin commented on workflow/spec/roadmap.md:
```diff
@@ -0,0 +1,42 @@
+
+## Short Term
+
+### Continue updating [specification document](spec.md)
```
This list and the next one, while all good things, seem relative minor in scope and kind of obvious clean-up things.  I think it might be good to include a list of architectural design points, or features, that you guys are considering to give people an idea of where the spec might change in the future.  Assuming you have some in mind of course.

tsurdilo replied:
yeah i just did this pr to get the conversation started. input welcome

cathyhongzhang commented on workflow/spec/changelog.md:
```diff
@@ -0,0 +1,22 @@
+- Time/Date/Duration parameter types to ISO-8601 format
+
+### Removed
+- End state (replaces with "end" property in other states)
```
Agree with Doug. People can go through the summary of the auto-generation, which provides more accurate and complete info on the change than a manually maintained list.

cathyhongzhang commented on workflow/spec/roadmap.md:
```diff
@@ -0,0 +1,42 @@
+
+
+## Short Term
+
```
Most of the short team work have been completed, thanks for Tihomir's contribution! 
Since the roadmap is an evolving target, I am thinking it may be better to divide the roadmap into time phases so that people have an idea of what's coming and in what time frame. We can update the spec to add new work items agreed by the community or move some work items to another time frame, etc.

tsurdilo replied:
ok sounds good. I will try to update and see what you guys think.

cathyhongzhang replied:
Thanks for coming up with these roadmap (they are very good and I will add my thought later). Since these "short term" or "long term" items are evolving, I am wondering if it makes sense to put these items as part of the spec documents since we need to keep obsoleting and updating them. I like Doug's suggestion. What we need is a Workflow primer which provides insight and in-depth explanations behind the design philosophy of the spec and the reasons/considerations for the definitions of the spec. Maybe we can track and update the short term and long term work items in the spec project meeting minutes, similar to what CloudEvents does. wdyt?

cathyhongzhang replied:
I can help write the primer later. Be patient with me since I have many other work to do:-)

salaboy commented:
/LGTM I think that we should merge this asap.

cathyhongzhang commented:
@tsurdilo Per your latest reply, you will update it based on our comments. I am still not seeing the updated one yet or am I missing something here?

tsurdilo commented:
@cathyhongzhang I was hoping that someone could help with generation of the changelog but it's fine. I updated the roadmap and also updated the generated changelog. Just FYI if anyone wants to maintain this sometimes in the future, the changelog was generated with https://github.com/github-changelog-generator/github-changelog-generator

tsurdilo commented:
@salaboy @cathyhongzhang @duglin updated. please check it again and leave your comments. thanks.

tsurdilo commented:
@cathyhongzhang @salaboy any info on the latest updates to this?

cathyhongzhang commented on workflow/spec/changelog.md:
```diff
@@ -0,0 +1,53 @@
+
+[Full Changelog](https://github.com/cncf/wg-serverless/compare/baef662dde4af54ae6a43490cf2e16692cff4708...HEAD)
+
+**Closed issues:**
```
Without checking into detail, it seems that what is in the auto-generated changelog does not quite match the Closed issues. Maybe I am wrong, but it seems that the Closed issues are manually added. If so, do we really want the extra work of manually maintaining a list, which could be error prone and not updated? An auto-generated changelog has all the info, right?

tsurdilo commented:
not sure what you mean. as i mentioned its generated. it would have to be regenerated each time its updated with the tool i mentioned in comment or some other you guys prefer

cathyhongzhang commented on workflow/spec/roadmap.md:
```diff
@@ -0,0 +1,26 @@
+* Create blueprint templates for different scenarios
+* Add Best Practices content
+* Add community content (presentations/blogs/videos)
+
```
Add more clarification/description for each definition. 

I feel that we might know well the semantics of the terminology/brief description of each definition in the spec, but a new comer/user might interpret them in an unexpected way. Adding examples will definitely help a lot. But adding more clarification will help too. The primer will also help people understand why we define or design it that way and they will have a base to challenge our design and contribute to the spec to make it better.

cathyhongzhang commented on workflow/spec/roadmap.md:
```diff
@@ -0,0 +1,26 @@
+## Long Term
+
+* Add Workflow Context (data)
+* Improve workflow error handling
```
Add synchronous and asynchronous workflow response options

We can discuss the exact scenarios and how do we support these scenarios.

cathyhongzhang commented:
> not sure what you mean. as i mentioned its generated. it would have to be regenerated each time its updated with the tool i mentioned in comment or some other you guys prefer

I am not sure if we need to document the changelog, which is a moving target and we have to keep updating it. I think people are interested in the spec and the design consideration behind each definition, not really all the change logs or closed issues imo. We can probably add a link that people can use to auto generate the change log in the primer document. wdyt?

tsurdilo commented:
@cathyhongzhang there is much value showing changes (as part of releases at least). Let's see what @duglin says as idk how cloudevents does this.

duglin commented:
I think it might depends on where you are in the overall process and state of the project. If you guys think you're still in the incubator phase where things are rapidly changing then a changelog might just be an annoyance and slow down your work. But, if you think the spec is at a stage where people are going to start to pick it up and implement it, then giving them a way to see what big changes have gone in recently would be nice. Asking them to look at the git logs isn't ideal. However, if the work is far enough along for a changelog then perhaps it's time to consider snap-shotting things - meaning creating releases.  Like a v0.1.

tsurdilo commented:
@duglin I think we are ready for a snapshot (0.1) once we resolve the outstanding open pull requests.

tsurdilo commented:
I'll close this pr and create a new one that just includes the roadmap doc. I think we really need to have this roadmap doc in order to start preparing for v0.1.
